draft-ietf-quic-recovery.md

```diff
@@ -461,21 +464,35 @@ The PTO period is the amount of time that a sender ought to wait for an
 acknowledgement of a sent packet.  This time period includes the estimated
 network roundtrip-time (smoothed_rtt), the variance in the estimate (4*rttvar),
 and max_ack_delay, to account for the maximum time by which a receiver might
-delay sending an acknowledgement.
+delay sending an acknowledgement.  When the PTO is armed for Initial or
+Handshake packet number spaces, the max_ack_delay is 0, as specified in
+13.2.5 of {{QUIC-TRANSPORT}}.
 
 The PTO value MUST be set to at least kGranularity, to avoid the timer expiring
 immediately.
 
+A sender computes its PTO timer every time an ack-eliciting packet is sent.
+When ack-eliciting packets are in-flight in multiple packet number spaces,
+the timer MUST be set for the packet number space with the earliest timeout,
+except for ApplicationData, which MUST be ignored until the handshake
+completes; see Section 4.1.1 of {{QUIC-TLS}}.  Not arming the PTO for
+ApplicationData prioritizes completing the handshake and prevents the server
+from sending a 1-RTT PTO packet before it has the keys to process a 1-RTT
+packet.  This simplifies to setting the PTO on the lowest active packet number
+space if packets from two different packet number spaces are sent when the PTO
+fires.
```

Comment:
I'm not sure if I understand this sentence. Consider the server that has just sent ServerHello and it's certificate, i.e. has outstanding Initial and Handshake data.
Isn't the whole point of this PR to arm the PTO timer based on both packet number spaces?

This is what I'd imagine would happen if no ACKs from the client are received:
t = 0: send Initial, immediately afterwards send Handshake. Set PTO timer for the Initial.
t = 1: PTO timer expires for the Initial packet. Send an Initial probe packet. Now the Handshake packet sent at t = 0 is the earliest outstanding packet, so set a PTO timer (with exponential backoff) for this packet.
t = 2: PTO timer expires for the Handshake packet. Send a Handshake probe packet. Now the Initial probe packet sent at t=1 is the earliest outstanding packet. The PTO timer for this packet would expire at t=5.

Is this correct?

Reply:
The Handshake PTO would expire at t=3, because of the backoff, which may have been what you meant.

The second Initial PTO would expire at t=5, as you said.

One note is that the text does recommend you send a packet in other packet number spaces when possible
"In addition to sending data in the packet number space for which the timer
expired, the sender SHOULD send ack-eliciting packets from other packet
number spaces with in-flight data, coalescing packets if possible."

Reply:
@marten-seemann is right, I missed this too. The text says that the PTO is set on the "lowest active PN space", which is incorrect. I would simply drop this sentence.

Reply:
Removed

Reply:
@ianswett No, I didn't mean t=3. The Handshake packet was sent sent at t=0, and we're doing one exponential backoff, meaning that the timer duration should be 2.
Or am I missing something here?

Reply:
I was thinking a Handshake packet was sent at 1, so it's my mistake.  Your example is correct given you're only going to send a single QUIC packet on PTO.